pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGeneratorFunctors.hpp

```diff
@@ -530,6 +530,386 @@ void applyNCGenGlobalPhase(
+        kokkos_swap(arr(i01), arr(i10));
+    };
+    applyNCGenerator2Functor<PrecisionT, decltype(core_function)>(
+        ExecutionSpace{}, arr_, num_qubits, controlled_wires, controlled_values,
+        wires, core_function);
+}
+
+template <class ExecutionSpace, class PrecisionT>
+void applyNCGenSingleExcitationPlus(
+    Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+    const std::size_t num_qubits,
+    const std::vector<std::size_t> &controlled_wires,
+    const std::vector<bool> &controlled_values,
+    const std::vector<std::size_t> &wires,
+    [[maybe_unused]] const bool inverse = false) {
+    auto core_function = KOKKOS_LAMBDA(
+        Kokkos::View<Kokkos::complex<PrecisionT> *> arr,
+        [[maybe_unused]] const std::size_t i00, const std::size_t i01,
+        const std::size_t i10, [[maybe_unused]] const std::size_t i11) {
+        arr(i00) *= -1;
+        arr(i01) *= Kokkos::complex<PrecisionT>{0.0, 1.0};
+        arr(i10) *= Kokkos::complex<PrecisionT>{0.0, -1.0};
+        arr(i11) *= -1;
+        kokkos_swap(arr(i01), arr(i10));
+    };
+    applyNCGenerator2Functor<PrecisionT, decltype(core_function)>(
+        ExecutionSpace{}, arr_, num_qubits, controlled_wires, controlled_values,
+        wires, core_function);
+}
+
+template <class PrecisionT, class FuncT> class applyNCGenerator4Functor {
+    using KokkosIntVector = Kokkos::View<std::size_t *>;
+
+    Kokkos::View<Kokkos::complex<PrecisionT> *> arr;
+    const FuncT core_function;
+    KokkosIntVector indices;
+    KokkosIntVector parity;
+    KokkosIntVector rev_wires;
+    KokkosIntVector rev_wire_shifts;
+    std::size_t mask{0U};
+    static constexpr std::size_t one{1U};
+    static constexpr std::size_t two{2U};
+    std::size_t i0011;
+    std::size_t i1100;
+
+  public:
+    template <class ExecutionSpace>
+    applyNCGenerator4Functor([[maybe_unused]] ExecutionSpace exec,
+                             Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+                             std::size_t num_qubits,
+                             const std::vector<std::size_t> &controlled_wires,
+                             const std::vector<bool> &controlled_values,
+                             const std::vector<std::size_t> &wires,
+                             FuncT core_function_)
+        : arr(arr_), core_function(core_function_) {
+        const std::size_t n_contr = controlled_wires.size();
+        const std::size_t n_wires = wires.size();
+        const std::size_t nw_tot = n_contr + n_wires;
+        PL_ASSERT(n_wires == 4);
+        PL_ASSERT(num_qubits >= nw_tot);
+
+        std::vector<std::size_t> all_wires;
+        all_wires.reserve(nw_tot);
+        all_wires.insert(all_wires.begin(), wires.begin(), wires.end());
+        all_wires.insert(all_wires.begin(), controlled_wires.begin(),
+                         controlled_wires.end());
+
+        std::tie(parity, rev_wires) =
+            reverseWires(num_qubits, wires, controlled_wires);
```

Comment:
This is a good candidate for a structured binding.
```suggestion
        auto [parity, rev_wires] = reverseWires(num_qubits, wires, controlled_wires);
```

Reply:
changed to use structured binding now, thanks!